src/lib/consensus/proof_of_stake.ml

```diff
@@ -1141,6 +1146,21 @@ module Make (Inputs : Inputs_intf) : Intf.S = struct
     in
     (protocol_state, consensus_transition_data)
 
+  let received_within_window (epoch, slot) ~time_received =
+    let open Time in
+    let time_received =
+      of_span_since_epoch (Span.of_ms (Unix_timestamp.to_int64 time_received))
+    in
+    let window_start = Epoch.slot_start_time epoch slot in
+    let window_end = add window_start Constants.network_window_length in
+    window_start < time_received && time_received < window_end
+
+  let is_valid consensus_state ~time_received =
+    let open Consensus_state in
+    received_within_window
```

Comment:
Why aren't they valid up to `expected_network_delay` after the slot ends?

Reply:
I checked Ourorobos, and it only describes the network delay as a multiplicative of slot time. We were encoding it differently and incorrectly before. This logic more closely matches what Ouroboros specifies.